### DISCUSSION TITLE
Adding string array to PrismaAdapter using AzureADProvider

### DISCUSSION THREAD

**Discussion by unknownsoldier08** [nextauthjs/next-auth#6501] [Help]
### Question 💬

I am using the PrismaAdapter with AzureADProvider for authentication and in my profile, I have a "roles" field that seems to come through just fine from the OAuthProfile.

My question is, how should I add this to my model for the PrismaAdapter? I've tried adding a roles model and referring to it from the User model, although this doesn't seem to work as I'd hope

```
[next-auth][debug][PROFILE_DATA] {
  OAuthProfile: {
    aud: '1d406fcd-eb11-4d68-bb2a-162ab020a366',
    iss: 'https://login.microsoftonline.com/f233c3a4-33df-4696-984e-5f9bf1cb0c09/v2.0',
    iat: 1674589484,
    nbf: 1674589484,
    exp: 1674593384,
    email: 'user@onmicrosoft.com',
    name: 'username',
    oid: 'e0aaebc7-3598-470a-9913-e985c475faa4',
    preferred_username: 'username@onmicrosoft.com',
    rh: '0.AVUApMMz8t8zlkaYTl-b8csMCc1vQB0R62hNuyoWKrAgo2aIAEk.',
    roles: [ 'Users' ],
    sub: 'EAqQBeueKBzh2DqBaKjG6XbwLQFvmGAgA6QJMK9IXSY',
    tid: 'f233c3a4-33df-4696-984e-5f9bf1cb0c09',
    uti: '50DmF6ch6UuBXiLeiwAyAA',
    ver: '2.0'
  }
```

> [next-auth][error][adapter_error_createUser] 
> https://next-auth.js.org/errors#adapter_error_createuser
> Invalid `p.user.create()` invocation in
> C:\Dev\marymede-alpha\node_modules\@next-auth\prisma-adapter\dist\index.js:6:38
> 
>   3 exports.PrismaAdapter = void 0;
>   4 function PrismaAdapter(p) {
>   5     return {
> → 6         createUser: (data) => p.user.create({
>               data: {
>                 name: 'Jay James Ciantar',
>                 email: 'jayc@z7n6t.onmicrosoft.com',
>                 image: null,
>                 roles: [
>                   'Users'
>                 ],
>                 ~~~~~~~~~
>                 emailVerified: null
>               }
>             })
> 
> Argument roles: Got invalid value
> [
>   'Users'
> ]
> on prisma.createOneUser. Provided List<String>, expected RolesCreateNestedManyWithoutUsersInput:
> type RolesCreateNestedManyWithoutUsersInput {
>   create?: RolesCreateWithoutUsersInput | List<RolesCreateWithoutUsersInput> | RolesUncheckedCreateWithoutUsersInput | List<RolesUncheckedCreateWithoutUsersInput>
>   connectOrCreate?: RolesCreateOrConnectWithoutUsersInput | List<RolesCreateOrConnectWithoutUsersInput>
>   connect?: RolesWhereUniqueInput | List<RolesWhereUniqueInput>
> }
> 

Any help would be greatly appreciated, thank you!

### How to reproduce ☕️

```ts
import NextAuth, { NextAuthOptions } from 'next-auth'
import AzureADProvider from 'next-auth/providers/azure-ad'
import { NextApiRequest, NextApiResponse } from 'next'
import { PrismaAdapter } from '@next-auth/prisma-adapter'
import prisma from 'lib/prisma'

export const authOptions: NextAuthOptions = {
  adapter: PrismaAdapter(prisma),
  providers: [
    AzureADProvider({
      clientId: process.env.AZURE_AD_CLIENT_ID,
      clientSecret: process.env.AZURE_AD_CLIENT_SECRET,
      tenantId: process.env.AZURE_AD_TENANT_ID,
      profile(profile) {
        return {
          id: profile.sub,
          name: profile.name,
          email: profile.email,
          image: profile.picture ?? null,
          roles: profile.roles ?? null,
        }
      },
    }),
  ],
  debug: true,
}

export default (req: NextApiRequest, res: NextApiResponse) => NextAuth(req, res, authOptions)
```

```ts
// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlserver"
  url      = env("DATABASE_URL")
}

model Account {
  id                String  @id @default(cuid())
  userId            String
  type              String
  provider          String
  providerAccountId String
  refresh_token     String? @db.Text
  access_token      String? @db.Text
  expires_at        Int?
  token_type        String?
  scope             String?
  id_token          String? @db.Text
  session_state     String?

  user User @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@unique([provider, providerAccountId])
}

model Session {
  id           String   @id @default(cuid())
  sessionToken String   @unique
  userId       String
  expires      DateTime
  user         User     @relation(fields: [userId], references: [id], onDelete: Cascade)
}

model User {
  id            String    @id @default(cuid())
  name          String?
  email         String?   @unique
  emailVerified DateTime?
  image         String?
  roles         Roles[]
  accounts      Account[]
  sessions      Session[]
}

model Roles {
  id    Int    @id @default(autoincrement())
  name  String @unique
  users User[]
}

model VerificationToken {
  identifier String
  token      String   @unique
  expires    DateTime

  @@unique([identifier, token])
}
```

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment]
You defined Roles as an object, but your provider returns it as a string. Unclear why you would need a Roles model though. You can [filter users](https://www.prisma.io/docs/concepts/components/prisma-client/filtering-and-sorting) on their role. So I would just save the role as-is on the user.

**Reply by unknownsoldier08** [reply]
In AzureADProvider, the OAuthProfile returns roles as an array because a user can be part of multiple roles.

How should I create my schema, if Roles returned for example,
```
roles: [
'Users',
'Administrators',
'Special1',
]
```
This was the reason that I created a roles object, but I am unsure how I should integrate it with the PrismaAdapter

**Comment by JohnnyTheTank** [comment]
hi @unknownsoldier08 I'm facing exactly the same problem. I want to get the roles from azure-ad gets assigned to the user or account. were you able to solve it?

**Comment by schmoli** [comment]
I'm working on this issue on my application right now, and I do at least have those roles assigned, all I put in my schema was:
```
roles String[]
```
with that said, AFAIK you can only assign a single role within Azure, so the direction I'm leaning is to just add `role String` to my schema and just grabbing the first if it exists. Unless there's a way to get multiple that I'm not seeing.

![CleanShot 2023-02-09 at 14 02 08](https://user-images.githubusercontent.com/11036865/217949500-05279fba-f06c-47b1-b55a-772eacf690b1.png)

**Reply by schmoli** [reply]
replying to myself, ok I did find out roles can be an array, you just have to add the user again with a different assignment (or they get one via group membership)
![CleanShot 2023-02-09 at 15 46 04](https://user-images.githubusercontent.com/11036865/217964354-6aff036b-d623-4faf-9d5a-bf71b4d2160a.png)

**Reply by JohnnyTheTank** [reply]
> I'm working on this issue on my application right now, and I do at least have those roles assigned, all I put in my schema was:
> 
> ```
> roles String[]
> ```
> 
> with that said, AFAIK you can only assign a single role within Azure, so the direction I'm leaning is to just add `role String` to my schema and just grabbing the first if it exists. Unless there's a way to get multiple that I'm not seeing.
> 
> ![CleanShot 2023-02-09 at 14 02 08](https://user-images.githubusercontent.com/11036865/217949500-05279fba-f06c-47b1-b55a-772eacf690b1.png)

to which model have you applied it? and what happens if you change the  role in azure?
